python/ray/data/impl/stats.py

```diff
@@ -0,0 +1,239 @@
+from typing import List, Optional, Dict, Set, Tuple
+import time
+import collections
+import numpy as np
+
+import ray
+from ray.data.block import BlockMetadata
+from ray.data.impl.block_list import BlockList
+
+
+def fmt(seconds: float) -> str:
+    if seconds > 1:
+        return str(round(seconds, 2)) + "s"
+    elif seconds > 0.001:
+        return str(round(seconds * 1000, 2)) + "ms"
+    else:
+        return str(round(seconds * 1000 * 1000, 2)) + "us"
+
+
+class _DatasetStatsBuilder:
+    """Helper class for building dataset stats.
+
+    When this class is created, we record the start time. When build() is
+    called with the final blocks of the new dataset, the time delta is
+    saved as part of the stats."""
+
+    def __init__(self, stage_name: str, parent: "DatasetStats"):
+        self.stage_name = stage_name
+        self.parent = parent
+        self.start_time = time.monotonic()
+
+    def build(self, final_blocks: BlockList) -> "DatasetStats":
+        stats = DatasetStats(
+            stages={self.stage_name: final_blocks.get_metadata()},
+            parent=self.parent)
+        stats.time_total_s = time.monotonic() - self.start_time
+        return stats
+
+
+@ray.remote(num_cpus=0)
+class _StatsActor:
+    """Actor holding stats for blocks created by LazyBlockList.
+
+    TODO(ekl) we should consider refactoring LazyBlockList so stats can be
+    extracted without using an out-of-band actor."""
+
+    def __init__(self):
+        self.start_time = time.time()
+        self.metadata = {}
+
+    def add(self, i, metadata):
+        self.metadata[i] = metadata
+        self.last_time = time.time()
+
+    def get(self):
+        return self.metadata, self.last_time - self.start_time
+
+
+class DatasetStats:
```

Comment:
I have a high level question: should we have per dataset stats or per operation (e.g. map) stats?

IIUC,for `dataset.map()`, the iterator time is recorded in the parent DatasetStats but the BlockExecStats is recorded in the child DatasetStats?

Should stats be operation centric:

```
dataset.map(..., stats: Optional[Stats])
print(stats)
```

Reply:
Generally each operation == a dataset.

> IIUC,for dataset.map(), the iterator time is recorded in the parent DatasetStats but the BlockExecStats is recorded in the child DatasetStats?

That's not the case. Iterator time is only relevant for the last dataset in a chain of transforms. Stats are always recorded for the child dataset.

> Should stats be operation centric:

I don't think this is usable since this makes stats recording very difficult for long pipelines.

Reply:
oh, I think I misunderstand your code.